src/fvtpf/tdyfvtpf.c

```diff
@@ -0,0 +1,1029 @@
+    *v_norm = PetscSqrtReal(norm_sum);
+  }
+
+  TDY_STOP_FUNCTION_TIMER()
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode TDyUpdateDiagnostics_FVTPF(void *context,
+                                          DM diags_dm,
+                                          Vec diags_vec) {
+  PetscErrorCode ierr;
+  PetscFunctionBegin;
+  TDY_START_FUNCTION_TIMER()
+  TDyFVTPF *fvtpf = context;
+  PetscInt c_start, c_end;
+  ierr = DMPlexGetHeightStratum(diags_dm,0,&c_start,&c_end); CHKERRQ(ierr);
+  PetscReal *v;
+  VecGetArray(diags_vec, &v);
+  for (PetscInt c = c_start; c < c_end; ++c) {
+    v[2*c+DIAG_LIQUID_SATURATION] = fvtpf->S[c];
+    v[2*c+DIAG_LIQUID_MASS] = fvtpf->rho[c] * fvtpf->V[c];
+  }
+  VecRestoreArray(diags_vec, &v);
+  TDY_STOP_FUNCTION_TIMER()
+  PetscFunctionReturn(0);
+}
+
+PetscErrorCode TDyUpdateState_Richards_FVTPF(void *context, DM dm,
+                                             EOS *eos,
+                                             MaterialProp *matprop,
+                                             CharacteristicCurves *cc,
+                                             PetscInt num_cells,
+                                             PetscReal *U) {
+  PetscErrorCode ierr;
+  PetscFunctionBegin;
+  TDyFVTPF *fvtpf = context;
+
+  PetscInt dim = 3;
+  PetscInt dim2 = dim*dim;
+  PetscInt cStart = 0, cEnd = num_cells;
+  PetscInt nc = cEnd - cStart;
+
+  // Compute the capillary pressure on all cells.
+  PetscReal Pc[nc];
+  for (PetscInt c=0;c<cEnd-cStart;c++) {
+    Pc[c] = fvtpf->Pref - U[c];
+    fvtpf->pressure[c] = U[c];
+  }
+
+  // Compute the saturation and its derivatives.
+  ierr = SaturationCompute(cc->saturation, fvtpf->Sr, Pc, fvtpf->S, fvtpf->dS_dP,
+                           fvtpf->d2S_dP2);
+
+  // Compute the effective saturation on cells.
+  PetscReal Se[nc];
+  for (PetscInt c=0;c<nc;c++) {
+    Se[c] = (fvtpf->S[c] - fvtpf->Sr[c])/(1.0 - fvtpf->Sr[c]);
+  }
+
+  // Compute the relative permeability and its derivative (w.r.t. Se).
+  ierr = RelativePermeabilityCompute(cc->rel_perm, Se, fvtpf->Kr, fvtpf->dKr_dS);
+
+  // Correct dKr/dS using the chain rule, and update the permeability.
+  for (PetscInt c=0;c<nc;c++) {
+    PetscReal dSe_dS = 1.0/(1.0 - fvtpf->Sr[c]);
+    fvtpf->dKr_dS[c] *= dSe_dS; // correct dKr/dS
+
+    for(PetscInt j=0; j<dim2; j++) {
+      fvtpf->K[c*dim2+j] = fvtpf->K0[c*dim2+j] * fvtpf->Kr[c];
+    }
+
+    // Also update water properties.
+    PetscReal P = fvtpf->Pref - Pc[c]; // pressure
+    ierr = EOSComputeWaterDensity(eos, P, &(fvtpf->rho[c]),
+                                  &(fvtpf->drho_dP[c]),
+                                  &(fvtpf->d2rho_dP2[c])); CHKERRQ(ierr);
+    ierr = EOSComputeWaterViscosity(eos, P, &(fvtpf->vis[c]),
+                                    &(fvtpf->dvis_dP[c]),
+                                    &(fvtpf->d2vis_dP2[c])); CHKERRQ(ierr);
+  }
+
+  PetscFunctionReturn(0);
+}
+
+static PetscErrorCode ComputeGeometry(TDyFVTPF *fvtpf, DM dm) {
```

Comment:
How does this function relate to the function you added to `TDyMesh` that computes geometry?